### PR TITLE
fix: right click popup should be closed on mousedown target element

### PR DESCRIPTION
**Pull request by rinick**
related issue:
https://github.com/ant-design/ant-design/issues/29853

I didn't check showAction or hideAction in isContextMenuOnly(), because when those 2 are defined, user might have very specific expectation about the mouse behavior, so I don't want to break existing code.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -405,7 +405,9 @@ export function generateTrigger(
       const root = this.getRootDomNode();
       const popupNode = this.getPopupDomNode();
       if (
-        !contains(root, target) &&
+        // mousedown on the target should also close popup when action is contextMenu.
+        // https://github.com/ant-design/ant-design/issues/29853
+        (!contains(root, target) || this.isContextMenuOnly()) &&
```
Comment (zombieJ) [inline src/index.tsx:410]: Could help add a test case for this? thx~

**Comment by ChocolateLoverRaj** [comment]
@rinick Can you link this PR to the issue it fixes by doing `Fixes #29853`?

File: src/index.tsx (hunk 1 above)
Comment (rinick) [reply src/index.tsx:410]: > Could help add a test case for this? thx~

I'm not familiar with enzyme
Just tried adding it in the end of basic.test.js -> 'contextMenu works' 

tried both `wrapper.trigger('mouseDown');` and  `wrapper.find('.target').simulate('mouseDown');`
neither close the popup for me.  

However, it does work in the storybook, So I'm not quite sure what to do.

Comment (zombieJ) [reply src/index.tsx:410]: You need simulate `contextMenu` first. And then simulate `mouseDown` next to check if it closed.

Comment (rinick) [reply src/index.tsx:410]: > You need simulate `contextMenu` first. And then simulate `mouseDown` next to check if it closed.

This is what I did, I tried both way to mouseDown but neither closes the popup
```
    it('contextMenu works', () => {
      const wrapper = mount(
        <Trigger
          action={['contextMenu']}
          popupAlign={placementAlignMap.left}
          popup={<strong>trigger</strong>}
        >
          <div className="target">contextMenu</div>
        </Trigger>,
      );

      wrapper.trigger('contextMenu');

      expect(wrapper.isHidden()).toBeFalsy();

      wrapper.trigger('mouseDown');
      wrapper.find('.target').simulate('mouseDown');
      expect(wrapper.isHidden()).toBeTruthy();
    });
```

Comment (rinick) [reply src/index.tsx:410]: @zombieJ simulate mouseDown on document is pretty tricky, this is the easiest work around I can find, please take a look

Comment (rinick) [reply src/index.tsx:410]: > You need simulate `contextMenu` first. And then simulate `mouseDown` next to check if it closed.

done

**Comment by zombieJ** [comment]
Seems contextMenu still not covered

截屏2021-04-13 上午11 52 03

**Comment by rinick** [comment]
> Seems contextMenu still not covered
> 
> 截屏2021-04-13 上午11 52 03

I can't see that page, hope it's fixed this time

**Comment by rinick** [comment]
@zombieJ ^

**Comment by rinick** [comment]
@zombieJ ^

**Comment by rinick** [comment]
@zombieJ ^

**Comment by zombieJ** [comment]
`+ rc-trigger@5.2.7`

**Pull request merged**